consensus/src/quorum_store/utils.rs

```diff
@@ -0,0 +1,308 @@
+        self.mempool_tx
+            .clone()
+            .try_send(msg)
+            .map_err(anyhow::Error::from)?;
+        // wait for response
+        match monitor!(
+            "pull_txn",
+            timeout(
+                Duration::from_millis(self.mempool_txn_pull_timeout_ms),
+                callback_rcv
+            )
+            .await
+        ) {
+            Err(_) => Err(anyhow::anyhow!(
+                "[direct_mempool_quorum_store] did not receive GetBatchResponse on time"
+            )),
+            Ok(resp) => match resp.map_err(anyhow::Error::from)?? {
+                QuorumStoreResponse::GetBatchResponse(txns) => Ok(txns),
+                _ => Err(anyhow::anyhow!(
+                    "[direct_mempool_quorum_store] did not receive expected GetBatchResponse"
+                )),
+            },
+        }
+    }
+}
+
+// TODO: unitest
+pub struct ProofQueue {
+    digest_queue: VecDeque<(HashValue, LogicalTime)>,
+    digest_proof: HashMap<HashValue, Option<ProofOfStore>>, // None means committed
+}
+
+impl ProofQueue {
+    pub(crate) fn new() -> Self {
+        Self {
+            digest_queue: VecDeque::new(),
+            digest_proof: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn push(&mut self, proof: ProofOfStore) {
+        match self.digest_proof.entry(*proof.digest()) {
+            Vacant(entry) => {
+                self.digest_queue
+                    .push_back((*proof.digest(), proof.expiration()));
+                entry.insert(Some(proof));
+            }
+            Occupied(mut entry) => {
+                if entry.get().is_some()
+                    && entry.get().as_ref().unwrap().expiration() < proof.expiration()
+                {
+                    entry.insert(Some(proof));
+                }
+            }
+        }
+    }
+
+    // gets excluded and iterates over the vector returning non excluded or expired entries.
+    // return the vector of pulled PoS, and the size of the remaining PoS
+    pub(crate) fn pull_proofs(
+        &mut self,
+        excluded_proofs: &HashSet<HashValue>,
+        current_time: LogicalTime,
+        max_txns: u64,
+        max_bytes: u64,
+    ) -> (Vec<ProofOfStore>, usize) {
+        let num_expired = self
+            .digest_queue
+            .iter()
+            .take_while(|(_, expiration_time)| *expiration_time < current_time)
+            .count();
+        for (digest, _expiration_time) in self.digest_queue.drain(0..num_expired) {
+            claims::assert_some!(self.digest_proof.remove(&digest));
+        }
+        debug!("QS: num_expired {}", num_expired);
+
+        let mut ret = Vec::new();
+        let mut cur_bytes = 0;
+        let mut cur_txns = 0;
+        let mut size = self.digest_queue.len();
+        for (digest, expiration) in self.digest_queue.iter() {
+            if *expiration >= current_time && !excluded_proofs.contains(digest) {
```

Comment:
From this code, if a proof is in excluded_proofs, it is no longer included in digest_queue. This doesn't seem the like the right behavior. If the block doesn't get executed, then the proof will be lost forever.

Reply:
This is ok because the transactions in the proof will be later pulled again from the mempool and get certified and committed.

Reply:
But is it more efficient to be able to retry with the same proof?